book/src/configuration.md

```diff
@@ -347,3 +347,58 @@ max-wrap = 25 # increase value to reduce forced mid-word wrapping
 max-indent-retain = 0
 wrap-indicator = ""  # set wrap-indicator to "" to hide it
 ```
+
+## Keymap config
+
+This section contains configuration that have to do with the behavior of some
+key bindings, but not with key bindings themselves. For keybindings, see [Key
+remapping].
+
+[Key remapping]: ./keymap.md
+
+### `[keys.supertab]` Section
+
+
+| Key        | Description | Default |
+|------------|-------------|---------|
+| `command` | If set to a command name, then when the cursor is in a position with non-whitespace to its left, instead of inserting a tab, it will run the given command. If there is only whitespace to the left, then it inserts a tab as normal. One can emulate a tabout behavior with this with the `move_parent_node_end` command. See [Tabout Example] | `None` |
+| `supercede_menu` | Normally, when a menu is on screen, such as when auto complete is triggered, the tab key is bound to cycling through the items. This means when menus are on screen, one cannot use the tab key to trigger the supertab command. If this option is set to true, the supertab command always takes precedence, which means one cannot use the tab key to cycle through menu items. One of the other bindings must be used instead, such as arrow keys or `C-n`/`C-p`. | `None` |
```

Comment:
Is it possible to move this under regular config, maybe as a `editor.keymap.supertab` or just `editor.supertab`? It looks like a lot of the changes in this PR are caused by the refactor for KeymapConfig. I think we should always prefer having multiple versions of a command to having config options affect a command and I'm worried that the KeymapConfig creates a precedence for adding more config options in the future.

I see it takes a MappableCommand which would be a blocker because it currently lives in helix-term (until #5581) but I think that taking a command name is unnecessary: it could be an enum of options for the different kind of motions implemented (`move_parent_node_end`). It seems like there would only be a few commands that you would ever want to configure for this option anyways.

With what I have in mind here, the `tab` in the default keymap would be `insert_tab` if we want tabout disabled by default and `supertab` if we want it enabled by default.

Reply:
> Is it possible to move this under regular config, maybe as a `editor.keymap.supertab` or just `editor.supertab`? It looks like a lot of the changes in this PR are caused by the refactor for KeymapConfig. I think we should always prefer having multiple versions of a command to having config options affect a command and I'm worried that the KeymapConfig creates a precedence for adding more config options in the future.

Hmm, so this is actually what I started with, but through discussion, it ended up changing to what it is now.

https://github.com/helix-editor/helix/pull/4443#discussion_r1192868909

When I think about the command set we have now and the precedence of having a separate command that has slightly different behavior, like `goto_line_end` vs `goto_line_end_newline`, or `change_selection` vs `change_selection_noyank`, the differences in behavior are isolated to themselves and I think generally make sense to have independent commands to enable both forms of behavior at once in one key map.

But in this case, the behavior of the feature of itself is not changed—it's kind of an odd case where it interacts with a UI element overriding the user's ability to invoke the feature. Having a `supertab` / `supertab_supercede_menu` command split seems inappropriate for the problem we're trying to solve, and also I'm not sure how it would even be implemented, since we'd need to know what command was invoked during the key event handling in order to know what the tab key should do when a menu is up.

> I see it takes a MappableCommand which would be a blocker because it currently lives in helix-term (until #5581) but I think that taking a command name is unnecessary: it could be an enum of options for the different kind of motions implemented (`move_parent_node_end`). It seems like there would only be a few commands that you would ever want to configure for this option anyways.

Hmm. The original goal of making it an arbitrary command name was for flexibility, both from the user's perspective, as well as any future uses of supertab, such as snippet placeholder cycling. We'd have to essentially keep an enum that would always be a subset of the command list, and any other command we want it to be able to do would have to require a code change.

That said, realistically, the concrete end goal was really just tabout. Some users might rebind it to something like `expand_selection`, `search_next`, `goto_definition` or perhaps others, but I can't say I believe that will be common.

If this is what we do though, is an enum what we want now? If `move_parent_node_end` is the only option, should it just be hardcoded until we add more options?

> With what I have in mind here, the `tab` in the default keymap would be `insert_tab` if we want tabout disabled by default and `supertab` if we want it enabled by default.

Well the idea of making tab = supertab was that if users want to enable it, they don't also have to rebind the tab key to supertab. Currently, if supertab is disabled, the supertab command just immediately calls the `insert_tab` command.